toolchain/lowering/lowering_context.cpp

```diff
@@ -109,4 +109,32 @@ auto LoweringContext::GetLoweredNodeAsType(SemanticsNodeId node_id)
   return type;
 }
 
+auto LoweringContext::GetLoweredNodeAsValue(SemanticsNodeId node_id)
+    -> llvm::Value* {
+  auto& node = lowered_nodes_[node_id.index];
+  if (node.is<llvm::Value*>()) {
+    return node.get<llvm::Value*>();
+  }
+  CARBON_CHECK(node.isNull())
+      << node_id << " is already set as a type, not a value";
+  // Empty values are built lazily.
+  // TODO: It might be better to built them at initialization, putting them in
+  // every IR even if not used. This is probably a performance decision since it
+  // would simplify this function.
```

Comment:
I'd guess the lazy approach is only a win for files that never need these. So if we end up actually avoiding them in most (all?) cases, then this would make sense. But if we don't then maybe not.

But it does seem really unfortunate to add this kind of logic here...

Would it make sense to have the semantics IR contain a node to build these when they're actually used, and then rely on that to produce them before their needed? And make any laziness happen at that layer rather than here?

Reply:
In order to do this with nodes, there would need to be a check that such nodes are only generated once before the first empty tuple/struct. Basically there'd need to be a bool tracking has_empty_tuple/struct on each SemanticsIR.

Instead, I can just add a bool tracking has_empty_tuple/struct on each SemanticsIR, set it true unconditionally whenever an empty tuple/struct is seen, and then special-case it at the start of lowering. I'll just go that route.

(not done yet, will get to it, but wanted to give you a chance to comment)

Reply:
I think that's also fine (and does avoid the awkwardly complex code here).

I still have a lingering question -- why are the empty case special / different from non-empty cases? Seems like you'd have to have a solution for general case to the problem you describe of only building the type once, and not sure why it wouldn't work for the empty case. But anyways, that shouldn't block this PR either way.

Reply:
Just a question of how much we want to optimize.

I was expecting empty tuples to be incredibly common because the language defines them as to-be-used instead of void. As a consequence, having them easy to compare against means code paths relating to empty tuples will be slightly faster.

If they're a separate type, that will function, but it means the common case will be constructing an empty tuple, and maybe doing a lookup to try to canonicalize it.

We can always remove special-casing and let the issue be addressed later. I think it just shifts costs though.

I'll leave this as-is since it seems the right solution isn't clear, and I do want this functioning.